### PR TITLE
[expo-cli][xdl][config] add expo publish --target managed|bare

### PR DESCRIPTION
**Pull request by esamelson**
# Why

https://www.notion.so/expo/App-entry-points-94433d55f7524837be7b37bccb2154a1

TL;DR In order to make `expo publish` able to create bundles for both the managed Expo client environment and bare native client environments, we need to be able to choose at runtime whether or not to resolve `.expo.*` files in the project source.

This PR adds a `--target` flag to `expo publish`; for `--target managed`, the packager resolves `.expo.*` files, and for `--target bare` the packager ignores them.

This PR also changes the default behavior for bare projects. Previously, `expo publish` **always** resolved `.expo.*` files, but now it only does so by default for managed workflow projects.

This is a **breaking change** for people who rely on `expo publish` resolving `.expo.*` files in bare workflow projects -- these developers will need to add the explicit `--target managed` flag to the command after this commit.

# How

- Added a new `getBareExtensions` to @expo/config to match `getManagedExtensions` and thread `target` param through to `startReactNativeServerAsync`
- Save the `target` in the packager info on disk; if there is an existing server instance running we only use it if the `target` matches, and close and restart otherwise
- Added `Project.isBareWorkflowProject` util for distinguishing between managed & bare projects

# Test Plan

Tested on a new bare workflow project from https://github.com/expo/expo/pull/7466
- [x] `--target` field overrides default
- [x] publishing with `--target managed` produces a build that errors with `The Expo SDK requires Expo to run...`
- [x] publishing with `--target bare` or unspecified target produces a build that runs
- [x] if `expo start` is running in a different process, publishing with `--target managed` uses the same server instance
- [x] if `expo start` is running in a different process, publishing with `--target bare` closes that instance and starts a new one

### CODE REVIEW

File: packages/xdl/src/Project.ts [hunk 1]
```diff
@@ -184,6 +186,30 @@ export async function currentStatus(projectDir: string): Promise<ProjectStatus>
   }
 }
 
+export async function isBareWorkflowProject(projectDir: string): Promise<boolean> {
```
Comment (brentvatne) [inline packages/xdl/src/Project.ts]:
```suggestion
export async function isBareWorkflowProjectAsync(projectDir: string): Promise<boolean> {
```

File: packages/expo-cli/src/commands/publish.ts [hunk 2]
```diff
@@ -40,16 +41,37 @@ export async function action(projectDir: string, options: Options = {}) {
       )}.`
     );
   }
+
+  const target =
+    options.target ?? ((await Project.isBareWorkflowProject(projectDir)) ? 'bare' : 'managed');
```
Comment (brentvatne) [inline packages/expo-cli/src/commands/publish.ts]:
```suggestion
    options.target ?? ((await Project.isBareWorkflowProjectAsync(projectDir)) ? 'bare' : 'managed');
```
